### PR TITLE
Add support for additional block number values in eth_getLogs

### LINKED ISSUE #8199: eth_getLogs cannot accept block strings other than "latest"
#### System information

Latest (devel branch)

Commit hash: 9c4797847265437d5dd1d00bdba23d549b5db2b1

#### Expected behaviour

Geth and other providers accept using `"finalized"` as a block identifier for the purposes of getting logs. Erigon also supports `"finalized"` for some block identifiers such as to `eth_getBlockByNumber`. This behavior should be ported correctly for the `eth_getLogs` RPC call.

As of writing, Geth, Reth, and Nethermind all support this behavior

#### Actual behaviour

Throws error: `"negative value for ToBlock: -4"`

#### Steps to reproduce the behaviour

`eth_getBlockByNumber` works okay:
```
curl \
-X POST \
-H 'Content-Type: application/json' \
--data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["finalized", false],"id":0}' \
<ERIGON_NODE_URL>
```
```
{"jsonrpc":"2.0","result": ... }
```

**`eth_getLogs` fails:**
```
curl \
-X POST \
-H 'Content-Type: application/json' \
--data '{"jsonrpc":"2.0","method":"eth_getLogs","params":[{"fromBlock":"finalized"}],"id":0}' \
<ERIGON_NODE_URL>
```
```
{"jsonrpc":"2.0","id":0,"error":{"code":-32000,"message":"negative value for FromBlock: -4"}}
```

#### More Info

The bug is [here](https://github.com/ledgerwatch/erigon/blob/devel/turbo/jsonrpc/eth_receipts.go#L122) where you exclude only `rpc.LatestBlockNumber` and do not consider-valid the other "magic int values" defined [here](https://github.com/ledgerwatch/erigon/blob/devel/rpc/types.go#L80)

### PR DESCRIPTION
**Pull request by pyropy**
Closes #8199

### CODE REVIEW

File: turbo/jsonrpc/eth_receipts.go [hunk 1]
```diff
@@ -91,10 +95,6 @@ func (api *APIImpl) GetLogs(ctx context.Context, crit filters.FilterCriteria) (t
 
 	if crit.BlockHash != nil {
 		num := rawdb.ReadHeaderNumber(tx, *crit.BlockHash)
-		//header, err := api._blockReader.HeaderByHash(ctx, tx, *crit.BlockHash)
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/eth_receipts.go:94]: Better use “blockReader” - it can reqd data from snapshots

**Comment by AskAlexSharov** [comment]
What is “magic block number”?

**Comment by pyropy** [comment]
> What is “magic block number”?

Haha just some name I got from the #8199 where it's referenced as "magic int values". I agree it's not the best, I'll change it :D

File: turbo/jsonrpc/eth_receipts.go (hunk 1 above)
Comment (pyropy) [reply turbo/jsonrpc/eth_receipts.go:94]: Should i use blockReader` for fetching all blocks, or just in case when `BlockHash` is given to me?

**Comment by pyropy** [comment]
@AskAlexSharov I've tested it out and seems to be working for pending, latest and earliest blocks. Just let me know if we need to support safe and finalized blocks too.

File: turbo/jsonrpc/eth_receipts.go (hunk 1 above)
Comment (AskAlexSharov) [reply turbo/jsonrpc/eth_receipts.go:94]: db storing only last 90K blocks - everything else is in snapshots

File: turbo/jsonrpc/eth_receipts.go [hunk 2]
```diff
@@ -225,6 +247,11 @@ func (api *APIImpl) GetLogs(ctx context.Context, crit filters.FilterCriteria) (t
 	return logs, nil
 }
 
+// getLogsIsValidBlockNumber checks if block number is valid integer or "latest", "pending", "earliest" block number
+func getLogsIsValidBlockNumber(blockNum *big.Int) bool {
+	return blockNum.IsInt64() && blockNum.Int64() >= -2
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/eth_receipts.go:252]: let's use some constant?

Comment (pyropy) [reply turbo/jsonrpc/eth_receipts.go:252]: Added change in #8370

**Pull request merged**